datalad/utils.py

```diff
@@ -760,8 +820,15 @@ def ensure_result_list(r):
     """
     return [r] if isinstance(r, dict) else ensure_list(r)
 
+@overload
+def ensure_list_from_str(s: str, sep: str='\n') -> Optional[list[str]]:
+    ...
 
-def ensure_list_from_str(s, sep='\n'):
+@overload
+def ensure_list_from_str(s: list[T], sep: str='\n') -> Optional[list[T]]:
+    ...
+
+def ensure_list_from_str(s: str | list[T], sep: str='\n') -> Optional[list[str]] | Optional[list[T]]:
```

Comment:
according to https://docs.python.org/3/library/typing.html#typing.overload type annotations of the actual implementation is ignored by type checking, thus IMHO it is better to not even bother type annotating it since it just mixes up more specific type annotations specified above in `@overload`'ed signatures and thus can somewhat misguide (and as you know I dislike duplication). Hence

```suggestion
def ensure_list_from_str(s, sep='\n'):
```

Reply:
I believe omitting the annotations on the actual implementation of the function would mean that mypy wouldn't type-check the body against the signature (in fact, the body wouldn't be type-checked at all).

Reply:
You are correct!

<details>
<summary>on this simple example based on the docs</summary> 

```python
from typing import overload,Union

@overload
def process(response: None) -> None:
    ...
@overload
def process(response: int) -> tuple[int, str]:
    ...
@overload
def process(response: bytes) -> str:
    ...
#def process(response:None|int|bytes) -> None|tuple[int,str]|str:
def process(response):
    # incompatible return e.g. in case of int or bytes
    return response
```
</details>

mypy   returns no errors, if I uncomment and use that line with combined signature - it picks it up but not because of overloads seems to me

<details>
<summary>on a more elaborate one where it is specifically violating overloads</summary> 

```python
from typing import overload,Union

@overload
def process(response: None) -> None:
    ...
@overload
def process(response: int) -> tuple[int, str]:
    ...
@overload
def process(response: bytes) -> str:
    ...
#def process(response:None|int|bytes) -> None|tuple[int,str]|str:
def process(response):
    # incompatible returns
    if isinstance(response, int):
        # specifically wrong for int -- must be tuple according to overload
        return str(int)
    elif isinstance(response, bytes):
        # specifically wrong for bytes
        return (1, str(bytes))
    # ok for remaining None
    return response


```
</details>

it seems to not bother finding problem in either case!  Seems like overloads are completely useless.

there is over 300 open issues mentioning `overload` https://github.com/python/mypy/search?q=overload&state=open&type=issues , so I think it is just mypy implementation is simply not there at all yet :-/ 

ok, let's proceed as is then, hopefully later mypy implementation would stick to what is written in the python docs and would simply ignore the annotation of the final implementation function and use overloads instead.